### PR TITLE
Add move tokens dialog

### LINKED ISSUE #1780: Add move funds between pots dialog
### Story

As a colony admin, I want to be able to move funds between domain pots, so that I have greater control over where funds are allocated.

### Description

This issue is the main event of the DLF/P phase 2 work! In this, we're adding the dialog component which allows selecting a "from" and "to" domain, as well as entering an amount and token. Submitting the dialog will dispatch the action created in #1779 to create the actual transaction.

Since we already have all the means to wire this thing up, we'll include that here too. The domain pot amounts for the selected token should be displayed below the domain dropdowns. If ETH is selected, then the USD conversion should also be shown (as in Figma).

Domains for which the current user does not have the effective funding role should be shown as greyed out in the domains dropdown, along with a tooltip. Form validation should be run to check that the selected from domain has enough funds to transfer the entered amount.

A "Move funds" button should be added to the admin tokens tab sidebar to open this dialog.

🌮 The dialog should accept a "to domain" as a param when opening, and should be opened with the currently selected domain preselected (or nothing selected if the user does not have permission.

### Design

Dialog: [Figma link](https://www.figma.com/file/L6scZfliLKnDt95nwan4jwzg/DLP-DLF?node-id=998%3A27269)

image

image

image

### This issue is complete when...

- [ ] Dialog exists with from/to dropdowns containing a list of domains, amount input, and token dropdown
  - [ ] Domains in dropdown are disabled if user does not have funding role to move from that pot (⚠️ need clarification on whether to domain also needs funding role)
  - [ ] Domain pot balances for selected token are displayed
  - [ ] Form is validated invalid if not enough funds in from pot
  - [ ] USD amount is displayed if ETH selected
- [ ] "Move funds" button added to colony admin tokens tab sidebar to open dialog
  - [ ] 🌮 Opens dialog with currently selected domain from tokens tab preselected as "to" domain

### Suggestions

* Getting whether the current user has funding permission for each domain shouldn't be too bad, but there might be some trickiness there, especially for greying out domain dropdown items.

### Subtasks

- [ ] ???

### PR DESCRIPTION
**Pull request by sprusr**
## Description

This PR adds a dialog which acts as a front for the `moveFundsBetweenPots` method on the colony contract.

image

**New stuff** ✨

* `TokensMoveDialog` and `TokensMoveDialogForm`
* `Select` core component now supports field status

**Changes** 🏗

* Ensure colony token balances are keyed by number and stored as `ImmutableMap` in `coloniesReducer`
* Refetch affected token balances after transaction in `colonyDomainsFetch` saga

**Deletions** ⚰️

* ???

## TODO

- [x] Wire up domain balance fetching (make sure the from domain actually has enough balance)
- [x] In `Tokens`, `canMoveTokens` should actually be as a result of some kind of check
- [x] Preselected `toDomain` when opening via "Move funds"?
- [x] Open `TokensMoveDialog` with selected domain as `toDomain` once #1799 is merged

Resolves #1780

### CODE REVIEW

File: src/modules/dashboard/reducers/coloniesReducer.ts [hunk 1]
```diff
@@ -13,6 +13,18 @@ import { createAddress, Address } from '~types/index';
 
 import { AllColoniesMap } from '../state/index';
 
+/**
+ * Convert a JS object of balances { 1: '...' } where keys may be number or
+ * string, to an ImmutableMap of balances keyed by numbers.
+ */
+const balancesMapFromObject = balances =>
```
Comment (ceolson01) [inline src/modules/dashboard/reducers/coloniesReducer.ts]: Nice 👍

File: src/modules/admin/components/Tokens/TokensMoveDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        fromDomain: undefined,
+        toDomain,
+        amount: '',
+        tokenAddress: ZERO_ADDRESS,
+      }}
+      validationSchema={validationSchema}
+      submit={ActionTypes.MOVE_FUNDS_BETWEEN_POTS}
+      error={ActionTypes.MOVE_FUNDS_BETWEEN_POTS_ERROR}
+      // Close dialog immediately to give way for GasStation
+      success={ActionTypes.MOVE_FUNDS_BETWEEN_POTS}
```
Comment (rdig) [inline src/modules/admin/components/Tokens/TokensMoveDialog.tsx]: We display no pending state whatsoever, right ? Like, it wasn't spec'd by product ?

Ie: The move funds dialog closes, the gas station opens (be we close it). Is there any way to tell there's a pending action ?

File: src/modules/admin/components/Tokens/TokensMoveDialogForm.tsx [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+  // Find the currently selected token
+  const [selectedTokenRef, selectedToken] = useMemo(
+    () => [
+      colonyTokenRefs.find(token => token.address === values.tokenAddress),
+      colonyTokens.find(token => token.address === values.tokenAddress),
+    ],
+    [colonyTokenRefs, colonyTokens, values.tokenAddress],
+  );
+
+  // Map the colony's tokens to Select options
```
Comment (rdig) [inline src/modules/admin/components/Tokens/TokensMoveDialogForm.tsx]: I don't know if this was spec'd or not, but shouldn't we sort the native token first ?

Screenshot from 2019-09-25 09-19-45

File: src/modules/admin/components/Tokens/TokensMoveDialogForm.tsx [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+              formattingOptions={{
+                delimiter: ',',
+                numeral: true,
+                numeralDecimalScale:
+                  (selectedToken && selectedToken.decimals) || 18,
+              }}
+            />
+          </div>
+          <div className={styles.tokenAmountSelect}>
+            <Select options={tokenOptions} name="tokenAddress" elementOnly />
```
Comment (rdig) [inline src/modules/admin/components/Tokens/TokensMoveDialogForm.tsx]: The select component should be made to open to the left. As it is, it adds a bottom scroll-bar, plus it cuts the options list:

Screenshot from 2019-09-25 09-19-57

File: src/modules/admin/components/Tokens/TokensMoveDialogForm.tsx [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    <>
+      <FormStatus status={status} />
+      <DialogSection>
+        <Heading
+          appearance={{ size: 'medium', margin: 'none' }}
+          text={MSG.title}
+        />
+      </DialogSection>
+      <DialogSection>
+        <Select options={domainOptions} label={MSG.from} name="fromDomain" />
```
Comment (rdig) [inline src/modules/admin/components/Tokens/TokensMoveDialogForm.tsx]: Can we better style the error message here? As it is it looks kinda broken...

Screenshot from 2019-09-25 09-17-22

Comment (sprusr) [reply src/modules/admin/components/Tokens/TokensMoveDialogForm.tsx]: Nice catch, thanks. Correctly set the `appearance.theme` now 👍

**Comment by sprusr** [comment]
Nice spot @rdig, I thought I had sorted the validation errors not going away but apparently not! Fixed now, and some other minor improvements to validation. Will merge after push and CI 👍

**Pull request merged**